SIPS/sip-23.md

```diff
@@ -0,0 +1,627 @@
+
+### Definitions
+
+#### Key
+
+A JSX key-like value, i.e., a `string`, `number`, or `null`.
+
+```typescript
+type Key = string | number | null;
+```
+
+#### SnapElement
+
+A rendered JSX element, i.e., an object with a `type`, `props`, and `key`.
+
+```typescript
+type SnapElement<Props = any> = {
+  type: string;
+  props: Props;
+  key: Key;
+};
+```
+
+#### SnapComponent
+
+A JSX component, i.e., a function which accepts `props`, and returns a
+`SnapElement`. All components MUST accept a `key` property in addition to the
+regular props.
+
+```typescript
+type SnapComponent<Props = Record<string, never>> =
+  (props: Props & { key: Key }) => SnapElement<Props>;
+```
+
+#### SnapNode
+
+A rendered JSX node, i.e., a `SnapElement`, `string`, `null`, or an array of
+`SnapNode`s. Note that HTML elements are not supported in Snap nodes.
+
+```typescript
+type SnapNode =
+  | SnapElement
+  | string
+  | null
+  | SnapNode[];
+```
+
+### Interface structure
+
+The Snap interface structure is defined using JSX components, and consists of:
+
+- A container element, `Container`, which wraps the entire interface.
+- A box element, `Box`, which contains the main content of the interface.
+- An optional footer element, `Footer`, which appears at the bottom of the
+  interface.
+
+![Snap interface structure](../assets/sip-23/interface-structure.png)
+
+#### Example
+
+Below is an example of a simple Snap interface structure using JSX components:
+
+```typescript jsx
+<Container>
+  <Box>
+    <Heading>My Snap</Heading>
+    <Text>Hello, world!</Text>
+  </Box>
+  <Footer>
+    <Button onClick={handleClick}>Click me</Button>
+  </Footer>
+</Container>
+```
+
+Each of the components used in the example is defined in the following sections.
+
+### Components
+
+#### Address
+
+The `Address` component renders a text element with an address. It accepts an
+`address` prop which is a string representing an Ethereum address.
+
+```typescript
+type AddressProps = {
+  address: `0x${string}`;
+};
+
+const Address: SnapComponent<AddressProps>;
+```
+
+##### Example
+
+```typescript jsx
+<Address address="0x4bbeEB066eD09B7AEd07bF39EEe0460DFa261520" />
+```
+
+#### Bold
+
+The `Bold` component renders a text element in bold. It accepts a `children`
+prop which can be a `string` or an array of `string`s. The component MUST be
+used inside a `Text` component.
+
+```typescript
+type BoldProps = {
+  children: string | string[];
+};
+
+const Bold: SnapComponent<BoldProps>;
+```
+
+##### Example
+
+```typescript jsx
+<Text>
+  Hello, <Bold>world</Bold>!
+</Text>
+```
+
+#### Box
+
+The `Box` component renders a box element. It accepts a `children` prop which
+can be a `SnapNode`. The component is used to group elements together.
+
+In addition, the component accepts an optional `alignment` prop which can be
+`'start'`, `'center'`, `'end'`, or `undefined`. The default value is
+`undefined`.
+
+```typescript
+type BoxProps = {
+  children: SnapNode;
```

Comment:
```suggestion
  children: SnapNode[];
```

Reply:
`SnapNode` accepts an array of `SnapNode`s itself.